### PR TITLE
Feature Multiple Distance Restraints

### PR DESCRIPTION
**Pull request by fjclark**
# Is this pull request to fix a bug, or to introduce new functionality?

This PR introduces functionality for setting up and running ABFE calculations with multiple distance restraints. This requires the functionality in [this](https://github.com/OpenBioSim/sire/pull/110) Sire PR.

## If this introduces new functionality...

Changes proposed in this pull request:

- Extend the Restraint class to hold multiple distance restraints and to write input files for calculations of this type with SOMD
- Implement a restraint search algorithm for multiple distance restraints. This currently:
    - Selects distance restraints for all heavy atoms in the ligand
    - Prioritises restraints between pairs of atoms with low SD of distance based on a simulation of the fully interacting complex
    - Promotes diversity of direction of the distance restraints by clustering candidate restraints by direction and selecting only one restraint per directional cluster
    - Chooses the flat-bottomed radius of the restraints to encompass 95 % of the distance sampled during the simulation of the fully interacting complex, so that the free energy of introducing the restraints is negligible

* I confirm that I have merged the latest version of `devel` into this branch before issuing this pull request (e.g. by running `git pull origin devel`): y
* I confirm that I have added a test for any new functionality in this pull request: y
* I confirm that I have added documentation (e.g. a new tutorial page or detailed guide) for any new functionality in this pull request: n (I will added a section to the ABFE tutorial if approved)
* I confirm that I have permission to release this code under the GPL3 license: y

When testing locally, all tests that I've added or that are related to functionality that I've changed pass. However, the following tests fail:
```
FAILED tests/Sandpit/Exscientia/Trajectory/test_trajectory.py::test_rmsd - AttributeError: 'TrajectoryIterator' object has no attribute 'rmsd'
FAILED tests/Sandpit/Exscientia/_SireWrappers/test_system.py::test_set_box - assert False
FAILED tests/Sandpit/Exscientia/_SireWrappers/test_system.py::test_rotate_box_vectors - AttributeError: 'TriclinicBox' object has no attribute 'rotate'
```

## Suggested reviewers:
@lohedges, @chryswoods

## Any additional context of information?

As  shown in the [Sire PR](https://github.com/OpenBioSim/sire/pull/110), so far my tests show no evidence of a significant difference compared to Boresch restraints. My test input scripts are [here](https://github.com/fjclark/bss_multiple_distance_restraints_example).

Thanks very much.

### CODE REVIEW

**Comment by lohedges** [comment]
Thanks, @fjclark:

> When testing locally, all tests that I've added or that are related to functionality that I've changed pass. However, the following tests fail:

Since the same tests passed during the CI here, I can only assume that locally you are testing against outdated versions of Sire and BioSimSpace? The `TriclinicBox` changes are fairly recent, so it's possible that you hadn't pulled them across to your fork.

**Comment by fjclark** [comment]
Yes, thanks - I had pulled the changes but forgotten to reinstall Sire.

**Comment by chryswoods** [comment]
Thanks - we are about to release 2023.4.0 (likely end of this week or early next week - it will depend on when we squash these last bugs).

Do you want this included in 2023.4.0? Or do you want more time to test it? I don't mind either way and am happy to review and merge this this week if you are happy for it to be included.

**Comment by lohedges** [comment]
I think @xiki-tempula might like to test this implementation soon, so I guess it might depend on whether he'd prefer that to be in 2023.4.0, or 2024.1.0.dev (or whatever it is).

**Comment by chryswoods** [comment]
2033.4.0 would mean we could further test and then fix bugs as part of 2023.4.1+. Going for 2023.5.0.dev would mean that it will be stuck in the dev package until we release 2023.5.0 (end of December).

The risk of going early is if this regresses or breaks existing functionality in any way. Do we have good regression tests for this code to demonstrate it will be safe to merge?

**Comment by xiki-tempula** [comment]
Thanks for informing me. I'm fine either way but I noticed that the implementation for the new restraint is for SOMD only and our pipeline currently only supports Gromacs, so I might not be testing it in the near future.

**Comment by fjclark** [comment]
> Thanks - we are about to release 2023.4.0 (likely end of this week or early next week - it will depend on when we squash these last bugs).
> 
> Do you want this included in 2023.4.0? Or do you want more time to test it? I don't mind either way and am happy to review and merge this this week if you are happy for it to be included.

Thanks very much. I think slightly more time to test it would be best. I'll also look into implementing this in Gromacs. I'll let you know once I've done this.

**Comment by xiki-tempula** [comment]
@fjclark Thanks

**Comment by chryswoods** [comment]
I reviewed and merged in the sire part of this so that it would make 2023.4.0 (sorry for not seeing the above first). The changes looked orthogonal to existing functionality, so relatively safe to include. I agree to wait for the next release for the BioSimSpace part - this will give you more time to test the interface before making this feature more public.

**Comment by lohedges** [comment]
@fjclark: Just following up on this:

> Thanks very much. I think slightly more time to test it would be best. I'll also look into implementing this in Gromacs. I'll let you know once I've done this.

Were you planning on doing this in updates to this PR, or as a separate one? If the SOMD implementation is ready then this could be merged (in case anyone wants to test it). Otherwise, I'm happy to wait for further updates, but will mark this as a WIP.

Cheers.

**Comment by fjclark** [comment]
> Were you planning on doing this in updates to this PR, or as a separate one? If the SOMD implementation is ready then this could be merged (in case anyone wants to test it). Otherwise, I'm happy to wait for further updates, but will mark this as a WIP.

Hi @lohedges, more testing has revealed some instabilities when the ligand is fully decoupled (even when only one restraint is active) which I didn't observed in my initial set of tests. I think it might be best to look into these more before merging, so could you please mark this as a work in progress? Thanks.

**Comment by fjclark** [comment]
Hi @lohedges, I've now finished implementing this for SOMD and for GROMACS and have validated both implementations by comparing with Boresch restraints. In both cases, I ran 5 independent replicates of the bound leg with each type of restraint and the errors are 95 % t-based confidence intervals:

**SOMD**
==============================================================================
**Boresch**

 

**Multiple Distance Restraints**

 

However, I observed a couple of crashes during the release restraint stage. To investigate these further, I ran 90 release restraint windows (10 replicates of 9 windows), and observed 10 crashes:

 

These crashes are sometimes observed when there are no restraints, and seem to occur much more frequently with open force fields compared to GAFF. As no crashes were observed with GROMACS, this seems to be a strange SOMD bug. More discussion is given [here](https://github.com/OpenBioSim/sire/issues/100).

For now, I've added a warning about potential instabilities whenever multiple distance restraints are used with SOMD.

**GROMACS**
==============================================================================
**Boresch**

 



**Multiple Distance Restraints**

 

**Overall**
==============================================================================
There is no evidence for a significant difference between Boresch restraints and multiple distance restraints for either engine. @lohedges  and @xiki-tempula, once you're happy with this PR I'll add a section to the tutorials showing how to use these. It's slightly different to Boresch restraints for GROMACS as a) The user has to specify `restraint-lambda` instead of `bonded-lambda` b) the calculation has to be split into two stages. This is because the only way I could see to do this was using a combination of bond restraint potentials (affected by `restraint-lambda`) and distance restraints (not affected by `restraint-lambda`).

**Comment by jmichel80** [comment]
That’s great work ! There seems to be a surprising difference in precision between the two engines, something to think about.

**Comment by lohedges** [comment]
Thanks @fjclark. Given the tests pass and this only touches sandpit functionality, I'd be happy to merge. Perhaps @xiki-tempula would like to check that it doesn't cause any issues with the internal Exs test suite before merging?

**Comment by xiki-tempula** [comment]
I have skim through the PR and I think it looks ok.

**Review by xiki-tempula — APPROVED** [review]

File: .github/workflows/Sandpit_exs.yml
Comment (xiki-tempula) [inline .github/workflows/Sandpit_exs.yml:38]: No need for two scikit-learn

Comment (fjclark) [reply .github/workflows/Sandpit_exs.yml]: Thanks for catching.

**Pull request merged**
